Add pattern to merge consecutive splits

devashishshankar opened the pull request:
Summary:
Fx pass for "split->split => split" pattern

{F959486105}

Test Plan:
* CI tests
* Run in a notebook on CMF MIMO

{F959938752}

Differential Revision: D45204109



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel reviewed (changes requested):
This implementation is a bit more complex than I would like.  Perhaps we could simplify it by breaking it into more reusable pieces.

Maybe something like:
```
class TorchSplit(CallFunction):
  """
  Matches a call to torch.split if it is in a normalized form.
  """
  def __init__(self, arg):
     # using KeywordArg("dim") for `dim` checks they all match 
     super().__init__(torch.split, arg, KeywordArg("dim"))
  
  def _match(...)
     m = super()._match(...)
     if not m:
         return m
     # check users are all unique getitems
     for user in node.users:
        m2 = CallFunction(operator.getitem, Arg(), Arg()).match(user)
        ...
        # check m2.args[0] == node and m2.args[1] is unique
     return m
```

Then a pattern like:
```
TorchSplit(CallFunction(operator.getitem, TorchSplit(Arg()), Arg())))
```

devashishshankar commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -79,4 +231,14 @@ def _split_cat_init():
         pattern = NormalizeSplit(pattern=pattern, extra_check=lambda arg: True)
-        pattern.register(patterns)
+        pattern.register(pass_1)
+    pattern_matcher_passes.append(pass_1)
```
You mean different variables in pre_grad.py? Like 

```
normalize_split_pass = PatternMatcherPass()
remove_consecutive_splits_pass = PatternMatcherPass()
```

jansel replied:
Yes, you can also just put the `PatternMatcherPass()` objects in this file and create a function here to run them called from `pre_grad.py`

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+        # check users are all unique getitems
+        seen_idxs = set()
+        for user in node.users:
+            if not match_node(user, {operator.getitem}):
```
This doesn't check that len(user.args)==2 and len(user.kwargs)==0.  You could check this with: `CallFunction(operator.getitem, Arg(), Arg()).match(user)`

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -82,4 +196,10 @@ def _split_cat_init():
+        pattern.register(normalize_split_pass)
+
+    # Pass 2: Merge consecutive splits
+    pass_2 = PatternMatcherPass()
```
Unused

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+    from .pre_grad import normalize_split_pass, remove_consecutive_splits_pass
 
+    # Pass 1: Normalize split cats
+    pass_1 = PatternMatcherPass()
```
Unused

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+    splits are unique getitems.
+    """
+
+    def __init__(self, arg):
+        # using KeywordArg("dim") for `dim` checks they all match
+        super().__init__(
+            torch.split, arg, Arg(), _users=MULTIPLE, dim=KeywordArg("dim")
```
```suggestion
    def __init__(self, arg, sizes):
        # using KeywordArg("dim") for `dim` checks they all match
        super().__init__(
            torch.split, arg, sizes, _users=MULTIPLE, dim=KeywordArg("dim")
```

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+        return m
+
+
+class MergeConsecutiveSplits(PatternEntry):
+    def apply(self, match: Match, graph: torch.fx.Graph, node: torch.fx.Node):
+        first_split = node.args[0].args[0]
+        first_split_sections = get_arg_value(first_split, 1, "split_size_or_sections")
+        next_split_sections = get_arg_value(node, 1, "split_size_or_sections")
+        next_split_index = node.args[0].args[1]
```
This might be a bit cleaner registering in global scope with something like:
```suggestion
@register_graph_pattern(
   TorchSplit(
       CallFunction(
           operator.getitem,
           TorchSplit(
               KeywordArg("first_split_input"),
               KeywordArg("first_split_sections"),
           )
        )
        KeywordArg("next_split_sections")
   ),
   pass_dict=remove_consecutive_splits_pass,
)
def merge_splits(match, first_split_input, first_split_sections, next_split_sections, dim):
   node = match.output_node()
```

Then you don't need to do stuff in `_split_cat_init`.

devashishshankar replied:
Good idea. Can I take this as a separate diff? I guess even split normalization can be done this way

jansel replied:
Change this one in this diff.  Update the other ones in another diff.

devashishshankar replied:
Oh, I didn't realize the decorator already existed. Changing :)

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+        ] = next_split_sections
+
+        first_split_input = first_split.args[0]
+        first_split_dim = get_arg_value(first_split, 2, "dim")
```
With the change above, you can just take many of these as args.

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+            if not match_node(user, {operator.getitem}):
+                # This should ideally never happen. Split user should always be a getitem
+                return FailedMatch(f"user of split not a getitem: {user}")
+            if user.args[1] in seen_idxs:
```
Should also check this is an integer.

devashishshankar replied:
What should I do if it is a non-int? I was assuming the 2nd input to getitem will always be an int. Since I don't handle non-ints, should I just return a FailedMatch for symint like cases?

jansel replied:
Just return `FailedMatch`

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+
+        to_remove = []
+
+        with graph.inserting_after(first_split_input):
```
Shouldn't this be before the first split node to preserve graph order?

devashishshankar replied:
Being after the split's input should also preserve graph order right? Since split can have only one tensor input?

jansel commented on torch/_inductor/fx_passes/split_cat.py:
```diff
@@ -48,20 +57,125 @@ def apply(self, match, graph, node):
+            new_split_num = 0
+            for split_num in range(len(first_split_sections)):
+                if split_num not in first_split_num_to_user:
+                    new_split_num += 1
```
Is this assuming the getitems are all in order and not having any gaps?

devashishshankar replied:
I am iterating over `first_split_sections` which is the sections of the split, not getitems. The condition `if split_num not in first_split_num_to_user` should only hit when there is a gap in getitems. But, let me add test cases explicitly verifying the gap and out-of-order cases

devashishshankar committed:
Add pattern to merge consecutive splits (#100107)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/100107

Fx pass for "split->split => split" pattern

{F959486105}

Test Plan:
* CI tests
* Run in a notebook on CMF MIMO

{F959938752}

Reviewed By: jansel

Differential Revision: D45204109

fbshipit-source-id: 3e29281cf52b27e4e4ecd1e471cea66cb18e6014
